File: app/src/main/java/com/dpis/module/settings/SystemFontScaleSettingsGateway.java
package com.dpis.module.settings;

import android.content.Context;
import android.provider.Settings;

public final class SystemFontScaleSettingsGateway {
    public Integer readPercent(Context context) {
        try {
            float scale = Settings.System.getFloat(
                    context.getContentResolver(),
                    Settings.System.FONT_SCALE,
                    SystemFontScaleToolState.scaleFromPercent(
                            SystemFontScaleToolState.DEFAULT_PERCENT));
            return SystemFontScaleToolState.percentFromScale(scale);
        } catch (RuntimeException e) {
            return null;
        }
    }

    public boolean canWrite(Context context) {
        return Settings.System.canWrite(context);
    }

    public boolean writePercent(Context context, int percent) {
        return Settings.System.putFloat(
                context.getContentResolver(),
                Settings.System.FONT_SCALE,
                SystemFontScaleToolState.scaleFromPercent(percent));
    }
}

File: app/src/main/java/com/dpis/module/tools/presentation/SystemFontScaleSettingsGateway.kt
package com.dpis.module.tools.presentation

import android.content.Context
import android.provider.Settings
import com.dpis.module.tools.SystemFontScaleToolPresenter
import com.dpis.module.tools.SystemFontScaleToolState

class SystemFontScaleSettingsGateway(
    private val context: Context,
) : SystemFontScaleToolPresenter.Gateway {
    override fun readPercent(): Int? = try {
        val scale = Settings.System.getFloat(
            context.contentResolver,
            Settings.System.FONT_SCALE,
            SystemFontScaleToolState.scaleFromPercent(
                SystemFontScaleToolState.DEFAULT_PERCENT,
            ),
        )
        SystemFontScaleToolState.percentFromScale(scale)
    } catch (_: RuntimeException) {
        null
    }

    override fun canWrite(): Boolean = Settings.System.canWrite(context)

    override fun writePercent(percent: Int): Boolean = Settings.System.putFloat(
        context.contentResolver,
        Settings.System.FONT_SCALE,
        SystemFontScaleToolState.scaleFromPercent(percent),
    )
}
